jupyterlab/handlers/ydoc_handler.py

```diff
@@ -0,0 +1,177 @@
+"""Echo WebSocket handler for real time collaboration with Yjs"""
+
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+import asyncio
+from typing import Optional
+
+from jupyter_server.base.handlers import JupyterHandler
+from jupyter_server.utils import ensure_async
+from jupyter_ydoc import ydocs as YDOCS
+from tornado import web
+from tornado.websocket import WebSocketHandler
+from ypy_websocket.websocket_server import WebsocketServer, YRoom
+
+YFILE = YDOCS["file"]
+
+RENAME_SESSION = 127
+
+
+class JupyterRoom(YRoom):
+    def __init__(self, type):
+        super().__init__(ready=False)
+        self.type = type
+        self.cleaner = None
+        self.watcher = None
+        self.document = YDOCS.get(type, YFILE)(self.ydoc)
+
+
+class JupyterWebsocketServer(WebsocketServer):
+    def get_room(self, path: str) -> JupyterRoom:
+        file_type, file_path = path.split(":", 1)
+        if file_path not in self.rooms.keys():
+            self.rooms[file_path] = JupyterRoom(file_type)
+        return self.rooms[file_path]
```

Comment:
I think I'm having an issue here. When opening a notebook as a plain text document closing it and opening the same notebook with the Notebook editor the content disappears.

I think the problem is that we are using the same key to store both rooms for two different types of files.
I believe we should use `type + path` as a key to store the room, this way we will create different rooms when opening the same document with different editors.

```suggestion
        file_type, file_path = path.split(":", 1)
        if path not in self.rooms.keys():
            self.rooms[path] = JupyterRoom(file_type)
        return self.rooms[path]
```

Reply:
Thanks Carlos, good catch!
I had never tried opening the same file with different editors. Now I'm seeing issues with the dirty flag "bouncing" between the windows where the file is opened as different types. I need to investigate.

Reply:
The issue is that when we have e.g. a notebook opened as such and also as a text document, modifying the notebook doesn't sync the text document, because they are actually two different (incompatible) Y documents. What happens is that when the notebook is changed, it is saved to disk after one second. Then the text document watches the file change and reverts to the file content, setting its dirty flag. After one second, it is saved to disk and the notebook watches the file change, and so on.
I think that for the long term fix, we need a kind of "Y bridge" between two different Y documents of the same file, so that we can sync the two in real time. But it is not going to be easy, if possible at all, because the text file is "unstructured" and it may be impossible to map character changes to higher-level notebook changes. Or maybe we could recreate the notebook from scratch, but then it would not be incremental updates.
For now, let's just break this circle by not actually saving if what we want to save is already on disk. The two opened documents will sync "through the file system", and not immediately, but I think it is fine.
What do you think @hbcarlos @fcollonval @jtpio @dmonad ?